### PR TITLE
Fix duplicate filter issues across all search tools

### PR DESCRIPTION
**Pull request by gokhanarkan**
# Fix duplicate filter issues across all search tools

Closes: #817

## **Problem**

Multiple search tools were incorrectly adding duplicate filters to user queries, causing GitHub API validation errors when users provided queries that already contained these filters. This affected three main areas:

### **1. Search Issues & Pull Requests (`search_issues`, `search_pull_requests`)**
- Added duplicate `is:` and `repo:` filters
- Caused "More than five AND/OR/NOT operators were used" errors
- Broke complex queries with existing filters

### **2. User & Organization Search (`search_users`, `search_orgs`)**  
- Added duplicate `type:` filters unconditionally
- Caused HTTP 422 validation errors for queries already containing `type:user` or `type:org`

**Example bugs:**
- **Issues**: `"repo:github/repo is:issue"` → `"is:issue repo:github/repo is:issue"` ❌
- **Users**: `"type:user location:seattle"` → `"type:user type:user location:seattle"` ❌

## **Solution**

Implemented comprehensive duplicate filter prevention across all search tools with a unified approach:

### **Core Helper Functions**
```go
// Detect specific filter:value combinations
func hasSpecificFilter(query, filterType, filterValue string) bool

// Detect any filter of a given type  
func hasFilter(query, filterType string) bool

// Convenience functions for common cases
func hasRepoFilter(query string) bool
func hasTypeFilter(query string) bool
```

### **Fixed Search Handlers**

**Issues/Pull Requests (`searchHandler`):**
```go
// Only add is:searchType if exact filter doesn't exist
if !hasSpecificFilter(query, "is", searchType) {
    query = fmt.Sprintf("is:%s %s", searchType, query)
}

// Only add repo filter if none exists  
if owner != "" && repo != "" && !hasRepoFilter(query) {
    query = fmt.Sprintf("repo:%s/%s %s", owner, repo, query)
}
```

**Users/Organizations (`userOrOrgHandler`):**
```go
// Only add type filter if none exists
if !hasTypeFilter(query) {
    searchQuery = "type:" + accountType + " " + query
}
```

### **Enhanced Debugging & Observability**

Added comprehensive logging to track filter logic and API interactions:
- **Server initialization**: Token validation and API endpoint logging
- **Query construction**: Shows original vs final queries with filter addition logic
- **API responses**: Detailed HTTP status, headers, and error bodies
- **Authentication tracking**: Masked token logging for security

This helps diagnose issues and validates that our fixes are working correctly in production.

## **What This Fixes**

✅ **`search_issues`**: No more duplicate `is:issue` or `repo:` filters  
✅ **`search_pull_requests`**: No more duplicate `is:pr` or `repo:` filters  
✅ **`search_users`**: No more duplicate `type:user` filters  
✅ **`search_orgs`**: No more duplicate `type:org` filters  
✅ **Complex queries**: OR operators and parentheses work without hitting API limits  
✅ **User experience**: Existing filters in queries are always preserved  
✅ **Debugging**: Clear logging shows exactly what's happening at each step

## **Testing**

Added extensive test coverage across all affected areas:

**Unit Tests:**
- **25+ test cases** for helper functions covering edge cases
- **Integration tests** for each search tool with duplicate filter scenarios  
- **Complex query tests** with OR operators and parentheses
- **Regression tests** ensuring existing functionality is preserved

**Manual Validation:**
- Verified fix works with authentication in place
- Tested against live GitHub API with various query patterns
- Confirmed logging provides clear visibility into filter logic

## **Real-World Impact**

This fix resolves the frustrating experience users had when their perfectly valid GitHub search queries would fail with cryptic API errors. Now users can:

- Use complex queries with confidence: `"repo:owner/repo is:issue (label:bug OR label:enhancement OR label:feature)"`
- Provide queries with existing filters without worrying about duplicates
- Get clear debug information when issues occur
- Have a consistent experience across all search tools

## **Alternatives Considered**

1. **API-level deduplication**: Let GitHub handle duplicate filters
   - **Discarded**: GitHub returns validation errors, poor UX

2. **Query rewriting**: Parse and reconstruct queries to remove conflicts  
   - **Discarded**: Too complex, fragile, could break edge cases

3. **Per-tool fixes**: Fix each search tool individually
   - **Discarded**: Code duplication, inconsistent behavior

4. **User education**: Document that users shouldn't include certain filters
   - **Discarded**: Bad UX, GitHub's own UI includes these filters

The chosen approach provides a clean, maintainable solution that works consistently across all search tools while preserving user intent and providing excellent debugging capabilities.

## **Migration Notes**

This is a backward-compatible fix. Existing queries will continue to work exactly as before, but queries that previously failed due to duplicate filters will now succeed.

No configuration changes or user action required - the fix is transparent and automatic.

### CODE REVIEW

**Comment by shreyabaid007** [comment]
Thanks @gokhanarkan for taking up [my issue #817](https://github.com/github/github-mcp-server/issues/817). This is a good approach to fixing the duplicate filter issue.

I know this is still in draft. However, I've identified that the exact same pattern as the original bug also exists in the [`userOrOrgHandler`](https://github.com/github/github-mcp-server/blob/main/pkg/github/search.go#L207) function and needs to be addressed as well. 

DUPLICATE FILTER ISSUE IN search.go
- **Location**: Line 207 in `pkg/github/search.go` - `userOrOrgHandler` function  
- **Link**: https://github.com/github/github-mcp-server/blob/main/pkg/github/search.go#L207

- **Issue**: Same pattern as original bug - unconditionally prepends `type:` filter  
- **Problematic Code**: 
```go
searchQuery := "type:" + accountType + " " + query
```

- **Impact**: If user query already contains `type:user` or `type:org`, creates duplicates and results in `HTTP 422 Validation Failed`

- **Total Affected Tools**: `search_users`, `search_pull_requests`, `search_users` and `search_orgs`

**Comment by gokhanarkan** [comment]
> Thanks @gokhanarkan for taking up [my issue #817](https://github.com/github/github-mcp-server/issues/817). This is a good approach to fixing the duplicate filter issue.
> 
> I know this is still in draft. However, I've identified that the exact same pattern as the original bug also exists in the [`userOrOrgHandler`](https://github.com/github/github-mcp-server/blob/main/pkg/github/search.go#L207) function and needs to be addressed as well.
> 
> DUPLICATE FILTER ISSUE IN search.go
> 
> * **Location**: Line 207 in `pkg/github/search.go` - `userOrOrgHandler` function
> * **Link**: https://github.com/github/github-mcp-server/blob/main/pkg/github/search.go#L207
> * **Issue**: Same pattern as original bug - unconditionally prepends `type:` filter
> * **Problematic Code**:
> 
> ```go
> searchQuery := "type:" + accountType + " " + query
> ```
> 
> * **Impact**: If user query already contains `type:user` or `type:org`, creates duplicates and results in `HTTP 422 Validation Failed`
> * **Total Affected Tools**: `search_users`, `search_pull_requests`, `search_users` and `search_orgs`

@shreyabaid007 I've fixed it using the same approach as the is: filter fix.

What I fixed:

- Added hasTypeFilter() helper function
- Modified userOrOrgHandler to only add type: when not already present
- Added comprehensive tests for search_users and search_orgs
- Extended search_pull_requests tests to prevent regressions

The fix:

```
// Before: always added type: filter (broken)
searchQuery := "type:" + accountType + " " + query

// After: only add if missing (fixed)
if !hasTypeFilter(query) {
    searchQuery = "type:" + accountType + " " + query
}
```

This prevents duplicate type:user and type:org filters that were causing HTTP 422 errors. Same pattern as your original issue, same solution!

Thanks for the thorough analysis - this will help a lot of users avoid those validation errors. 🚀

File: pkg/github/search_utils_test.go [hunk 1]
```diff
@@ -0,0 +1,326 @@
...
+			query:      "repo:github/github-mcp-server is:issue (label:critical OR label:urgent)",
+			filterType: "is",
+			expected:   true,
+		},
+		{
+			name:       "query with complex OR expression checking repo",
+			query:      "repo:github/github-mcp-server is:issue (label:critical OR label:urgent)",
+			filterType: "repo",
+			expected:   true,
+		},
```
Comment (LuluBeatson) [inline pkg/github/search_utils_test.go:93]: Edge case where filter is preceded by non-whitespace
```suggestion
		},
		{
			name:       "label in parentheses",
			query:      "(label:bug OR owner:bob) is:issue",
			filterType: "label",
			expected:   true,
		},
```

**Comment by gokhanarkan** [comment]
> Awesome work!
> 
> One small thing: the regex currently expects the filter to be preceded by whitespace. So `hasFilter` will fail on `query="(label:bug OR owner:bob) is:issue"`, `filterType="label"`. Unlikely to occur but perhaps worth fixing now so it doesn't fail quietly later.
> 
> Otherwise approach is great and I've tested the tools
> 
> Example: Search Pull Requests

@LuluBeatson Updated both regex patterns to handle this edge case!

**Comment by shreyabaid007** [comment]
Great work! 🚀
